### DISCUSSION TITLE
Run Supabase Locally With Sign In

### DISCUSSION THREAD

**Discussion by accmltr** [supabase/supabase#17447] [Questions] [answered]
Hi, 
I'm trying to write tests for my Supabase database functions and policies. However, my policies and functions do a lot of work dependant on the ID of the user making the API calls to determine whether they have ownership of rows they try to edit in other tables. They use the built auth.uid() database function for this.

Unfortunately when running Supabase locally and running my tests, all the auth.* functions are unavailable...

So my question is, is there a way to sign in on a fake user account for tests?

Here is an example function:
```
declare
    result text;
begin
    -- Check if there is space for the character
    if character_cap_reached(auth.uid()) then
        raise exception 'User already owns maximum capacity of characters. Characters owned: %.', (select count(*) from public.player_characters where user_id = auth.uid())::text;
    end if;

    -- Check if the character name already exists
    if exists (select * from public.player_characters where name = character_name) then
        raise exception 'Character name already in use: %.', character_name;
    end if;

    -- Insert the new character
    insert into public.player_characters (user_id, name)
    values (auth.uid(), character_name);

    -- Return the new UUID as text
    return '';
end;
```

Here is an example test on that function:
```
BEGIN;
SELECT plan(2);

-- Test 1: Your test assertion here
SELECT is(true, true, 'Your test description');

-- Test 2: Your test assertion here
SELECT create_player_character('Brick Eater');
SELECT is(true, (
    SELECT count(*) = 1 
    FROM player_characters 
    WHERE name = 'Brick Eater'
    ), 'Brick Eater should have been created in the player_characters table.');


ROLLBACK;
```

Please, any help on this would be greatly appreciated.

Kind regards,
Johan

**Comment by GaryAustin1** [comment, marked as answer]
I do something like this to test RLS...
auth.uid() and auth.jwt() work.  You can provide other claims as needed.

```SQL
set session role authenticated;
set request.jwt.claims to '{"role":"authenticated", "sub":"5950b438-b07c-4012-8190-6ce79e4bd8e5"}';

explain analyze SELECT count(*) FROM rlstest;
set session role postgres;
```

**Reply by accmltr** [reply]
Thank you so much, really appreciate it!!

**Comment by accmltr** [comment]
I also just found this, which also has the answer if anyone wants to read up more:
https://usebasejump.com/blog/testing-on-supabase-with-pgtap#installing
